src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,272 @@
+    this.retrier = retrier;
+  }
+
+  private ExecutionBlockingStub executionBlockingStub() {
+    return ExecutionGrpc.newBlockingStub(channel)
+        .withInterceptors(TracingMetadataUtils.attachMetadataFromContextInterceptor())
+        .withCallCredentials(callCredentialsProvider.getCallCredentials())
+        .withDeadlineAfter(remoteOptions.remoteTimeout.getSeconds(), TimeUnit.SECONDS);
+  }
+
+  private static class Execution {
+    private final ExecuteRequest request;
+    private final OperationObserver observer;
+    private final RemoteRetrier retrier;
+    private final CallCredentialsProvider callCredentialsProvider;
+    private final ProgressiveBackoff backoff;
+    private final Supplier<ExecutionBlockingStub> executionBlockingStubSupplier;
+
+    private Operation operation;
+
+    Execution(ExecuteRequest request,
+        OperationObserver observer,
+        RemoteRetrier retrier,
+        CallCredentialsProvider callCredentialsProvider,
+        Supplier<ExecutionBlockingStub> executionBlockingStubSupplier) {
+      this.request = request;
+      this.observer = observer;
+      this.retrier = retrier;
+      this.callCredentialsProvider = callCredentialsProvider;
+      this.backoff = new ProgressiveBackoff(this.retrier::newBackoff);
+      this.executionBlockingStubSupplier = executionBlockingStubSupplier;
+    }
+
+    ExecuteResponse start() throws IOException, InterruptedException {
+      // Execute has two components: the Execute call and (optionally) the WaitExecution call.
+      // This is the simple flow without any errors:
+      //
+      // - A call to Execute returns streamed updates on an Operation object.
+      // - We wait until the Operation is finished.
+      //
+      // Error possibilities:
+      // - An Execute call may fail with a retriable error (raise a StatusRuntimeException).
+      //   - If the failure occurred before the first Operation is returned, we retry the call.
+      //   - Otherwise, we call WaitExecution on the Operation.
+      // - A WaitExecution call may fail with a retriable error (raise a StatusRuntimeException).
+      //   In that case, we retry the WaitExecution call on the same operation object.
+      // - A WaitExecution call may fail with a NOT_FOUND error (raise a StatusRuntimeException).
+      //   That means the Operation was lost on the server, and we will retry to Execute.
+      // - Any call can return an Operation object with an error status in the result. Such
+      //   Operations are completed and failed; however, some of these errors may be retriable.
+      //   These errors should trigger a retry of the Execute call, resulting in a new Operation.
+      Preconditions.checkState(operation == null);
+
+      ExecuteResponse response = null;
+      while (response == null) {
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
```

Comment:
In GrpcRemoteExecutor, we use refreshIfUnauthenticated outside of retry block: if there is a unauthenticated error, retrier will stop retrying and propagate the error to refreshIfUnauthenticated which will call retrier again. It will reset the retry time counter so we could retry more than `--remote_retry` times which is not expected.

I will review all the refreshIfUnauthenticated call sites and refactor the code as per https://github.com/bazelbuild/bazel/pull/12156#issuecomment-701104900.

Reply:
Thanks! I think it's actually better this way, but I didn't know if this was intentional or not and I was not quite sure what it does. I also appreciate the new comment!